### PR TITLE
Implement debugpy for python

### PR DESCRIPTION
**Pull request by tonybaloney**
Implements #1035 

- [x] The purpose of this PR is explained in this message or in an issue. If an issue please include a reference as \#\<issue\_number\>.
- [x] Tests are included and/or updated for code changes.
- [x] Proper license headers are included in each file.

### CODE REVIEW

File: src/startupscriptgenerator/src/python/scriptgenerator.go [hunk 1]
```diff
@@ -321,6 +336,23 @@ func (gen *PythonStartupScriptGenerator) buildPtvsdCommandForModule(moduleAndArg
...
 
+func (gen *PythonStartupScriptGenerator) buildDebugPyCommandForModule(moduleAndArgs string, appDir string) string {
+	waitarg := ""
+	if gen.DebugWait {
+		waitarg = "--wait-for-client"
+	}
+
+	cdcmd := ""
+	if appDir != "" {
+		cdcmd = fmt.Sprintf("cd %s && ", appDir)
```
Comment (asw101) [inline src/startupscriptgenerator/src/python/scriptgenerator.go:347]: Will this handle paths with spaces?

**Comment by tonybaloney** [comment]
@karthiknadig please can you check the debug py command in https://github.com/microsoft/Oryx/pull/1036/files#diff-a47503bf32b2396a61d0f2ab7307019d0a6d4c9bd92b29dcf203f19cf275e73fR350 looks ok

File: src/startupscriptgenerator/src/python/scriptgenerator.go [hunk 2]
```diff
@@ -321,6 +336,23 @@ func (gen *PythonStartupScriptGenerator) buildPtvsdCommandForModule(moduleAndArg
...
+	}
+
+	pycmd := fmt.Sprintf("%spython -m debupy --listen %s:%s %s -m %s",
```
Comment (karthiknadig) [inline src/startupscriptgenerator/src/python/scriptgenerator.go:350]: This looks right.

File: src/startupscriptgenerator/src/python/scriptgenerator.go (hunk 1 above)
Comment (tonybaloney) [reply src/startupscriptgenerator/src/python/scriptgenerator.go:347]: I didn't see any shell escaping in any of the other examples. probably out of scope for this PR, but worth asking the maintainers.

**Comment by tonybaloney** [comment]
@microsoft/oryx-contrib can someone review this PR please

File: images/runtime/python/template.Dockerfile [hunk 3]
```diff
@@ -39,6 +39,7 @@ RUN ${IMAGES_DIR}/runtime/python/install-dependencies.sh
 RUN pip install --upgrade pip \
     && pip install gunicorn \
     && pip install ptvsd \
```
Comment (arroyc) [inline images/runtime/python/template.Dockerfile:41]: we dont need ptvsd anymore i guess if we are adding debugpy

File: src/startupscriptgenerator/src/python/main.go [hunk 4]
```diff
@@ -34,7 +34,7 @@ func main() {
 		" e.g. 'application:app'.")
 	defaultAppDebugModulePtr := scriptCommand.String("defaultAppDebugModule", "application.py", "Module to run if "+
 		"running the app in debug mode, e.g. 'application.py start_dev_server'. Has no effect if -debugAdapter isn't used.")
-	debugAdapterPtr := scriptCommand.String("debugAdapter", "", "Python debugger adapter. Currently, only 'ptvsd' is "+
+	debugAdapterPtr := scriptCommand.String("debugAdapter", "", "Python debugger adapter. Currently, only 'ptvsd' and 'debugpy' are "+
```
Comment (arroyc) [inline src/startupscriptgenerator/src/python/main.go:37]: remove ptvsd

File: src/startupscriptgenerator/src/python/main.go [hunk 5]
```diff
@@ -123,6 +123,7 @@ func main() {
 		scriptBuilder.WriteString("pip install --upgrade pip\n")
 		scriptBuilder.WriteString("pip install gunicorn\n")
 		scriptBuilder.WriteString("pip install ptvsd\n")
```
Comment (arroyc) [inline src/startupscriptgenerator/src/python/main.go:125]: remove

File: src/startupscriptgenerator/src/python/scriptgenerator.go [hunk 6]
```diff
@@ -34,12 +34,15 @@ type PythonStartupScriptGenerator struct {
...
 }
 
-const SupportedDebugAdapter = "ptvsd" // Not using an array since there's only one at the moment
 const GeneratingCommandMessage = "Generating `%s` command for '%s'"
 
 const DefaultHost = "0.0.0.0"
 const DefaultBindPort = "80"
 
+func getSupportedDebugAdapters() []string {
+	return []string{"ptvsd", "debugpy"}
```
Comment (arroyc) [inline src/startupscriptgenerator/src/python/scriptgenerator.go:43]: remove ptvsd?

File: src/startupscriptgenerator/src/python/scriptgenerator.go [hunk 7]
```diff
@@ -113,8 +116,14 @@ func (gen *PythonStartupScriptGenerator) GenerateEntrypointScript() string {
 		if appModule != "" {
 			if gen.shouldStartAppInDebugMode() {
 				logger.LogInformation("Generating debug command for appDebugModule.")
-				println(fmt.Sprintf(GeneratingCommandMessage, "ptvsd", appDebugModule))
-				command = gen.buildPtvsdCommandForModule(appDebugModule, appDirectory)
+				println(fmt.Sprintf(GeneratingCommandMessage, gen.DebugAdapter, appDebugModule))
+				switch gen.DebugAdapter {
+				case "ptvsd":
+					command = gen.buildPtvsdCommandForModule(appDebugModule, appDirectory)
```
Comment (arroyc) [inline src/startupscriptgenerator/src/python/scriptgenerator.go:122]: remove ptvsd as its deprecated?

**Review by arroyc — CHANGES_REQUESTED** [review]
I made some comments .. hopefully you can address those .. sorry about late response ..  currently we are short on people .. please expect some delay ... if you have urgent request please email

**Comment by tonybaloney** [comment]
@arroyc updates completed

**Review by arroyc — COMMENTED** [review]
this looks good now .. what i will do i will create a PR from this branch .. because external user's  pr do not trigger validation builds .. and without a successful build we cannot merge it to master

File: src/startupscriptgenerator/src/python/scriptgenerator.go [hunk 8]
```diff
@@ -111,10 +114,19 @@ func (gen *PythonStartupScriptGenerator) GenerateEntrypointScript() string {
 		}
 
 		if appModule != "" {
+			// Patch all legacy ptvsd debug adaptor calls to debugpy
+			if gen.DebugAdapter == "ptvsd" {
```
Comment (qianz2) [inline src/startupscriptgenerator/src/python/scriptgenerator.go:118]: 👍

**Comment by qianz2** [comment]
> this looks good now .. what i will do i will create a PR from this branch .. because external user's pr do not trigger validation builds .. and without a successful build we cannot merge it to master

Hey Arjun, I'm using the last commit id from this PR to manually trigger a validation build, it somehow worked. So let's wait until it passed the build, then we should be able to merge it, you don't need to create another branch for it.

**Review by qianz2 — APPROVED** [review]

**Comment by arroyc** [comment]
> > this looks good now .. what i will do i will create a PR from this branch .. because external user's pr do not trigger validation builds .. and without a successful build we cannot merge it to master
> 
> Hey Arjun, I'm using the last commit id from this PR to manually trigger a validation build, it somehow worked. So let's wait until it passed the build, then we should be able to merge it, you don't need to create another branch for it.

It didn't work, so we have to create a separate branch to merge it in.

**Comment by arroyc** [comment]
closing because of #1050

**Pull request closed without merging**